runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1031,64 +1379,81 @@ static iree_status_t iree_hal_hip_device_queue_alloca(
     iree_hal_allocator_pool_t pool, iree_hal_buffer_params_t params,
     iree_device_size_t allocation_size,
     iree_hal_buffer_t** IREE_RESTRICT out_buffer) {
+  *out_buffer = NULL;
+  IREE_TRACE_ZONE_BEGIN(z0);
+
   iree_hal_hip_device_t* device = iree_hal_hip_device_cast(base_device);
-  IREE_RETURN_IF_ERROR(
-      iree_hal_hip_set_context(device->hip_symbols, device->hip_context));
+  uint64_t queue_affinity_mask =
+      ((iree_hal_queue_affinity_t)1 << device->topology.count);
+  queue_affinity_mask = queue_affinity_mask | (queue_affinity_mask - 1);
+  queue_affinity &= queue_affinity_mask;
+
+  int device_ordinal = iree_math_count_trailing_zeros_u64(queue_affinity);
+  queue_affinity = (uint64_t)1 << device_ordinal;
 
-  if (device->supports_memory_pools &&
-      !iree_all_bits_set(params.type, IREE_HAL_MEMORY_TYPE_HOST_VISIBLE)) {
+  iree_status_t status = iree_ok_status();
+  if (!iree_all_bits_set(params.type, IREE_HAL_MEMORY_TYPE_HOST_VISIBLE) &&
+      (device->supports_memory_pools ||
+       iree_hal_hip_allocator_isa(iree_hal_device_allocator(base_device)))) {
     iree_hal_buffer_t* buffer = NULL;
 
-    IREE_RETURN_IF_ERROR(iree_hal_hip_memory_pools_prepare_buffer(
-        &device->memory_pools, device->hip_dispatch_stream, pool, params,
-        allocation_size, &buffer));
+    status = iree_hal_hip_device_pepare_async_alloc(device, params,
+                                                    allocation_size, &buffer);
 
-    iree_status_t status = iree_hal_deferred_work_queue_enqueue_alloc(
-        device->work_queue, wait_semaphore_list, signal_semaphore_list, buffer);
+    iree_hal_hip_device_semaphore_buffer_operation_callback_data_t*
+        callback_data = NULL;
     if (iree_status_is_ok(status)) {
-      status = iree_hal_deferred_work_queue_issue(device->work_queue);
+      status = iree_hal_hip_device_make_buffer_callback_data(
+          device, device->host_allocator, queue_affinity, wait_semaphore_list,
+          signal_semaphore_list, buffer,
+          IREE_HAL_HIP_DEVICE_SEMAPHORE_OPERATION_ASYNC_ALLOC, &callback_data);
     }
-    if (iree_status_is_ok(status)) {
+    if (iree_status_is_ok(status) && wait_semaphore_list.count == 0) {
+      status = iree_hal_hip_dispatch_thread_add_dispatch(
+          device->topology.devices[device_ordinal].dispatch_thread,
+          &iree_hal_hip_device_perform_buffer_operation_now, callback_data);
       *out_buffer = buffer;
-    } else {
-      iree_hal_hip_buffer_set_allocation_empty(buffer);
-      iree_hal_resource_release(&buffer->resource);
+      IREE_TRACE_ZONE_END(z0);
```

Comment:
avoid early exits like this - I think this is a bug (as nearly all early exits introduce bugs, if not extremely careful :) - the check in the if is for status ok **and** wait count 0, which is odd, and then the code above is allocating but nothing here cleans up if status is a failure. I think you wanted to early exit if the wait count was 0 but that should happen earlier, I think. or not - but it's not possible to tell now :)

Reply:
I can fix the early exit here, but if wait count == 0, then we can't rely on the semaphores to unblock the job for us and we have to kick it off ourselves, otherwise if we have a wait semaphore, we have to add the callbacks to the semaphores.